dbms/src/Flash/Mpp/GRPCReceiveQueue.h

```diff
@@ -0,0 +1,250 @@
+private:
+    void * tag;
+};
+
+using GRPCReceiveKickFunc = std::function<grpc_call_error(KickReceiveTag *, grpc_call *)>;
+using AsyncRetryConnection = std::pair<KickReceiveTag *, grpc_call *>;
+
+// This is a FIFO data struct.
+//
+// When MPMCQueue is full, put the information related with AsyncRequestHandler
+// into this queue so that we can kick it into GRPCCompletionQueue when the
+// MPMCQueue is not full and try to re-write the data.
+class AsyncRequestHandlerWaitQueue
+{
+public:
+    bool empty()
+    {
+        std::lock_guard lock(mu);
+        return wait_retry_queue.empty();
+    }
+
+    void push(AsyncRetryConnection conn)
+    {
+        std::lock_guard lock(mu);
+        wait_retry_queue.push(conn);
+    }
+
+    AsyncRetryConnection pop()
+    {
+        std::lock_guard lock(mu);
+        if (wait_retry_queue.empty())
+            return AsyncRetryConnection(nullptr, nullptr);
+
+        AsyncRetryConnection ret_conn = wait_retry_queue.front();
+        wait_retry_queue.pop();
+
+        return ret_conn;
+    }
+
+    std::queue<AsyncRetryConnection> popAll()
+    {
+        std::lock_guard lock(mu);
+        if (wait_retry_queue.empty())
+            return std::queue<AsyncRetryConnection>{};
+
+        std::queue<AsyncRetryConnection> ret_conn = std::move(wait_retry_queue);
+        return ret_conn;
+    }
+
+private:
+    std::mutex mu;
+    std::queue<AsyncRetryConnection> wait_retry_queue;
+};
+
+using AsyncRequestHandlerWaitQueuePtr = std::shared_ptr<AsyncRequestHandlerWaitQueue>;
+
+enum class GRPCReceiveQueueRes
+{
+    OK,
+    FINISHED,
+    FULL,
+    CANCELLED,
+};
+
+// When MPMCQueue is full, thread that pushes data into MPMCQueue will be blocked which
+// is unexpected when that's a grpc thread. Because tiflash's grpc threads responsible for
+// receiving data are limited and the available grpc threads will be less and less when more
+// and more push operation are blocked. In order to avoid the above case, we create this
+// non-blocking queue so that push operation will not be blocked when the queue is full.
+template <typename T>
+class GRPCReceiveQueue
+{
+private:
+    using MsgChannelPtr = std::shared_ptr<ConcurrentIOQueue<T>>;
+
+public:
+    GRPCReceiveQueue(MsgChannelPtr recv_queue_, AsyncRequestHandlerWaitQueuePtr conn_wait_queue_, const LoggerPtr & log_)
+        : recv_queue(std::move(recv_queue_))
+        , conn_wait_queue(std::move(conn_wait_queue_))
+        , log(log_)
+    {}
+
+    // For gtest usage.
+    GRPCReceiveQueue(MsgChannelPtr & recv_queue_, GRPCReceiveKickFunc func)
+        : recv_queue(recv_queue_)
+        , log(Logger::get())
+        , kick_func(func)
+    {}
+
+    const String & getCancelReason() const
+    {
+        return recv_queue->getCancelReason();
+    }
+
+    // Cancel the send queue, and set the cancel reason
+    bool cancelWith(const String & reason)
+    {
+        auto ret = recv_queue->cancelWith(reason);
+        if (ret)
+            handleRemainingTags();
```

Comment:
It seems that for cancel, the remaining messages can not be handled.

Reply:
> It seems that for cancel, the remaining messages can not be handled.

Not for processing the remaining messages, but for awaking the remaining `AsyncRequestHandler` to exit and notify the `ExchangeReceiver` that a connection has been closed, or the receiver will hang up.